### ISSUE TITLE
[Bug] 2.0 Resizing leaves black background

### ISSUE THREAD

**Issue by camgaertner** [wangwenx190/framelesshelper#122]
Hi thanks for your work on this project! Upon integrating the 2.0 version, I noticed this behavior when resizing the frameless window. Indeed, the jitter is fixed, but the window now has a black background appearing quickly in the previous space the window occupied. The previous version didn't seem to have this behavior.

https://user-images.githubusercontent.com/10566658/165296112-feb8f653-220e-4093-8f01-f2f1680fd39d.mp4

Additional Information:
```
Graphics Backend:
qputenv("QT_OPENGL", "angle");
qputenv("QT_ANGLE_PLATFORM", "d3d9");

Windows:
Windows 10 Pro
19044.1645

Graphics Card / Driver:
NVIDIA GTX 1080 Ti
Driver: 27.21.14.66.11
```

**Comment by wangwenx190** [comment]
Looks like your are demenstrating a Qt Quick application. For Qt Quick applications, the whole user interface is rendered through 3D graphics API such as d3d11, opengl, vulkan and metal. The black background you observed is caused by the corresponding graphics API, not FramelessHelper. It will behave differently if you switch to a different graphics API. This issue is not fixable from the FramelessHelper side, and I also think it's beyond the ability of a frameless helper. If the previous 1.x version doesn't have this issue, it may be because the new implementation doesn't remove the whole window frame so Windows's internal rendering logic is different with before. The 1.x implementation will remove the whole window frame and thus causes flicker and jitter during window resizing, but it seems OpenGL can be more stable, the new implementation get rid the flicker and jitter by preserving the window frame, so looks like we can't have both advantages.

**Comment by wangwenx190** [comment]
By the way, can you provide your Qt version? Thanks!

**Comment by camgaertner** [comment]
> By the way, can you provide your Qt version? Thanks!

This is on Qt 6.3.0.

>For Qt Quick applications, the whole user interface is rendered through 3D graphics API such as d3d11, opengl, vulkan and metal. The black background you observed is caused by the corresponding graphics API, not FramelessHelper. It will behave differently if you switch to a different graphics API.

For 6.0 it looks like everything goes through QRhi (https://www.qt.io/blog/graphics-in-qt-6.0-qrhi-qt-quick-qt-quick-3d). I think I'll be forced to use the default backend (rhi) for other reasons, so using the software backend isn't an option unfortunately. However, after I disabled the frameless helper, I just noticed the black bars remain during resizing, so looks like a Qt or graphics backend issue as you suggested. I noticed there is less jitter when resizing though (both with and without framelesshelper have some jitter though).

Have you tested this library on Qt 6+? I would be curious if you had any thoughts on the bit of jitter that's remaining.

**Comment by wangwenx190** [comment]
I always use the latest stable version of Qt to do developing, so currently I'm using Qt 6.3.0 as well. But the information you provided makes me a little confused. You said you are using Qt6, however, you also set the `QT_OPENGL` and `QT_ANGLE_PLATFORM` environment variables. So looks like you want to use the ANGLE backend instead of the Desktop OpenGL. But this won't work in Qt6 because all ANGLE related code paths in Qt have been removed and the only choice is the Desktop OpenGL. But these two environment variables will work as expected in Qt5. And the issue you observed is very specific to the graphics card drivers. For example, I've tested FramelessHelper on Intel and AMD graphics cards and both of them don't have this issue at all. But you are using NVIDIA graphics cards so it may be a NVIDIA only issue, I guess. And I'm developing and testing on Windows 11, it may also have different result compared to Windows 10. This issue is quite complicated and to be honest fixing this kind of issues is really beyond the ability of a frameless helper.

> I noticed there is less jitter when resizing though (both with and without framelesshelper have some jitter though).

This is because the most important code to reduce jitter have been merged into Qt itself sometime ago. The patch was contributed by me. It's merged after Qt 6.2.1 is released, so the first version that has the workaround is Qt 6.2.2. Qt 6.3 of course have that code as well, so the jitter is much more reduced for all Qt-based applications. The remaining jitter is because Qt's drawing system have some rounding issue internally, and it's not fixable from outside of Qt. I've seen several patches to improve the situation merged into QtBase and QtDeclarative sometime ago, maybe when Qt 6.4 is released, the jitter can be reduced more. Due to it's caused by internal Qt logic, we can't do anything to improve the situation.

**Comment by wangwenx190** [comment]
Wait, I watched the screen recording you uploaded here for multiple times, and after some deeper thoughts, I think you are not using the OpenGL backend at all. From what I see from the video, the application should be using the Direct3D 11 backend. If you are using the OpenGL backend, there should be two differences: one is there will be a strange black bar on top of your homemade title bar, the other one is the black background during window resizing should not be visible. You should re-check your Qt RHI backend.

You can change your Qt RHI backend by calling `QQuickWindow::setGraphicsApi()` before any Qt Quick elements are created.

**Comment by camgaertner** [comment]
> You said you are using Qt6, however, you also set the QT_OPENGL and QT_ANGLE_PLATFORM environment variables. So looks like you want to use the ANGLE backend instead of the Desktop OpenGL. But this won't work in Qt6 because all ANGLE related code paths in Qt have been removed and the only choice is the Desktop OpenGL. But these two environment variables will work as expected in Qt5.

Good point, I think this is just left over code from Qt5.

> And the issue you observed is very specific to the graphics card drivers. For example, I've tested FramelessHelper on Intel and AMD graphics cards and both of them don't have this issue at all. But you are using NVIDIA graphics cards so it may be a NVIDIA only issue, I guess. And I'm developing and testing on Windows 11, it may also have different result compared to Windows 10. This issue is quite complicated and to be honest fixing this kind of issues is really beyond the ability of a frameless helper.

That's fair.

> This is because the most important code to reduce jitter have been merged into Qt itself sometime ago. The patch was contributed by me. It's merged after Qt 6.2.1 is released, so the first version that has the workaround is Qt 6.2.2. Qt 6.3 of course have that code as well, so the jitter is much more reduced for all Qt-based applications. The remaining jitter is because Qt's drawing system have some rounding issue internally, and it's not fixable from outside of Qt. I've seen several patches to improve the situation merged into QtBase and QtDeclarative sometime ago, maybe when Qt 6.4 is released, the jitter can be reduced more. Due to it's caused by internal Qt logic, we can't do anything to improve the situation.

Understand. Thanks for that work as well, I can tell an improvement using Qt 6! Looking forward to seeing what you've done in Qt 6.4. 

Thanks for your response, I'll close this issue since it's out of scope.
